File: client/src/repositories/dashboardRepository.ts
import { clientRepository } from './clientRepository';
import { taskRepository } from './taskRepository';
import { defaultTaskSort } from '@/utils/taskSort';
import { isOverdue } from '@/utils/formatters';
import { DashboardSummary } from '@/types';

function rate(part: number, total: number): number {
  if (total === 0) return 0;
  return Math.round((part / total) * 100);
}

/** Same shape the old server /api/dashboard/summary returned — now computed
 * locally so the Home page works offline. */
async function summary(): Promise<DashboardSummary> {
  const [clients, tasks] = await Promise.all([
    clientRepository.list({ sort: 'createdAt', order: 'desc' }),
    taskRepository.list(),
  ]);

  const totalClients = clients.length;
  const completedClients = clients.filter((c) => c.status === 'completed').length;
  const pendingClients = clients.filter((c) => c.status !== 'completed').length;

  const totalTasks = tasks.length;
  const completedTasks = tasks.filter((t) => t.status === 'completed').length;
  const pendingTasks = tasks.filter((t) => t.status === 'pending').length;
  const inProgressTasks = tasks.filter((t) => t.status === 'in-progress').length;
  const overdueTasks = tasks.filter((t) => isOverdue(t.dueDate, t.status)).length;

  const recentClients = clients.slice(0, 5);
  const recentTasks = [...tasks].sort(defaultTaskSort).slice(0, 5);

  return {
    clients: {
      total: totalClients,
      completed: completedClients,
      pending: pendingClients,
      completionRate: rate(completedClients, totalClients),
    },
    tasks: {
      total: totalTasks,
      completed: completedTasks,
      pending: pendingTasks,
      inProgress: inProgressTasks,
      overdue: overdueTasks,
      completionRate: rate(completedTasks, totalTasks),
    },
    recentClients,
    recentTasks,
  };
}

export const dashboardRepository = { summary };

File: flutter_app/lib/data/repositories/dashboard_repository.dart
import '../../models/enums.dart';
import '../../models/models.dart';
import '../../utils/formatters.dart';
import 'client_repository.dart';
import 'task_repository.dart';

int _rate(int part, int total) => total == 0 ? 0 : ((part / total) * 100).round();

/// The Home counters, computed locally so the page works offline. Same shape
/// the old server /api/dashboard/summary returned.
class DashboardRepository {
  const DashboardRepository(this._clients, this._tasks);

  final ClientRepository _clients;
  final TaskRepository _tasks;

  DashboardSummary summary() {
    final clients = _clients.list(const ClientQuery(sort: 'createdAt'));
    final tasks = _tasks.list();

    final completedClients = clients.where((c) => c.status == EntityStatus.completed).length;
    final completedTasks = tasks.where((t) => t.status == EntityStatus.completed).length;

    // Completed clients linger on Home for a day, then drop out of the
    // "recent" list — they remain fully manageable on the Clientes screen.
    final recentClients = clients
        .where((c) => !isHiddenFromHome(
              completed: c.status == EntityStatus.completed,
              completedAt: c.completedAt,
            ))
        .take(5)
        .toList();

    // Tasks leave Home the moment they are checked off: Home is meant to show
    // what is still open. A finished task lives out its last 24h on the
    // Tarefas screen (see TaskRepository.purgeExpiredCompleted) and is then
    // deleted for good.
    final recentTasks = (tasks
          ..sort(defaultTaskSort))
        .where((t) => t.status != EntityStatus.completed)
        .take(5)
        .toList();

    return DashboardSummary(
      clients: DashboardCounts(
        total: clients.length,
        completed: completedClients,
        pending: clients.where((c) => c.status != EntityStatus.completed).length,
        completionRate: _rate(completedClients, clients.length),
      ),
      tasks: DashboardCounts(
        total: tasks.length,
        completed: completedTasks,
        pending: tasks.where((t) => t.status == EntityStatus.pending).length,
        inProgress: tasks.where((t) => t.status == EntityStatus.inProgress).length,
        overdue: tasks.where((t) => isOverdue(t.dueDate, t.status)).length,
        completionRate: _rate(completedTasks, tasks.length),
      ),
      recentClients: recentClients,
      recentTasks: recentTasks,
    );
  }
}
